File: pyoptix/context.py
import weakref
import atexit
from pyoptix._driver import NativeContextWrapper
from pyoptix.enums import ExceptionType
from pyoptix.mixins.scoped import ScopedObject


class Context(ScopedObject):
    def __init__(self):
        self._program_cache = {}
        self._ray_gen_programs = {}
        self._exception_programs = {}
        self._miss_programs = {}
        self._destroyables = []

        self._native = NativeContextWrapper()
        ScopedObject.__init__(self, self._native)

        push_context(self)

    def __del__(self):
        self._mark_all_objects_destroyed()

    def _mark_all_objects_destroyed(self):
        for destroyable in self._destroyables:
            if destroyable() is not None:
                destroyable().mark_destroyed()

    @property
    def program_cache(self):
        return self._program_cache

    def push(self):
        if current_context() == self:
            raise RuntimeError("Cannot push: Context is already at the top of the stack")
        else:
            push_context(self)

    def pop(self):
        if current_context() == self:
            pop_context()
        else:
            raise RuntimeError("Cannot pop: Context is not at the top of the stack")

    def launch(self, entry_point_index, width, height=None, depth=None):
        if not height:
            self._native.launch_1d(entry_point_index, width)
        elif not depth:
            self._native.launch_2d(entry_point_index, width, height)
        else:
            self._native.launch_3d(entry_point_index, width, height, depth)

    def _create_accelerator(self, builder, traverser):
        obj = self._native.create_accelerator(builder, traverser)
        self._destroyables.append(weakref.ref(obj))
        return obj

    def _create_geometry(self):
        obj = self._native.create_geometry()
        self._destroyables.append(weakref.ref(obj))
        return obj

    def _create_buffer(self, buffer_type):
        obj = self._native.create_buffer(buffer_type)
        self._destroyables.append(weakref.ref(obj))
        return obj

    def _create_geometry_group(self):
        obj = self._native.create_geometry_group()
        self._destroyables.append(weakref.ref(obj))
        return obj

    def _create_geometry_instance(self):
        obj = self._native.create_geometry_instance()
        self._destroyables.append(weakref.ref(obj))
        return obj

    def _create_group(self):
        obj = self._native.create_group()
        self._destroyables.append(weakref.ref(obj))
        return obj

    def _create_material(self):
        obj = self._native.create_material()
        self._destroyables.append(weakref.ref(obj))
        return obj

    def _create_program_from_file(self, file_name, function_name):
        obj = self._native.create_program_from_file(file_name, function_name)
        self._destroyables.append(weakref.ref(obj))
        return obj

    def _create_selector(self):
        obj = self._native.create_selector()
        self._destroyables.append(weakref.ref(obj))
        return obj

    def _create_texture_sampler(self):
        obj = self._native.create_texture_sampler()
        self._destroyables.append(weakref.ref(obj))
        return obj

    def _create_transform(self):
        obj = self._native.create_transform()
        self._destroyables.append(weakref.ref(obj))
        return obj

    def get_ray_type_count(self):
        return self._native.get_ray_type_count()

    def set_ray_type_count(self, ray_type_count):
        if not isinstance(ray_type_count, int) or ray_type_count < 0:
            raise TypeError('Index offset must be a positive integer')

        self._native.set_ray_type_count(ray_type_count)

    def get_entry_point_count(self):
        return self._native.get_entry_point_count()

    def set_entry_point_count(self, entry_point_count):
        if not isinstance(entry_point_count, int) or entry_point_count < 0:
            raise TypeError('Entry point count must be a positive integer')

        self._native.set_entry_point_count(entry_point_count)

    def get_cpu_num_of_threads(self):
        return self._native.get_cpu_num_of_threads()

    def set_cpu_num_of_threads(self, cpu_num_of_threads):
        if not isinstance(cpu_num_of_threads, int) or cpu_num_of_threads < 0:
            raise TypeError('Number of threads must be a positive integer')

        self._native.set_cpu_num_of_threads(cpu_num_of_threads)

    def get_stack_size(self):
        return self._native.get_stack_size()

    def set_stack_size(self, stack_size_bytes):
        if not isinstance(stack_size_bytes, int) or stack_size_bytes < 0:
            raise TypeError('Stack size is the number of bytes that must be a positive integer')

        self._native.set_stack_size(stack_size_bytes)

    def get_available_devices_count(self):
        return self._native.get_available_devices_count()

    def get_device_name(self, device_id):
        return self._native.get_device_name(device_id)

    def get_device_compute_capability(self, device_id):
        return self._native.get_device_compute_capability(device_id)

    def get_enabled_device_count(self):
        return self._native.get_enabled_device_count()

    def get_enabled_devices(self):
        return self._native.get_enabled_devices()

    def set_devices(self, devices):
        self._native.set_devices(devices)

    def get_used_host_memory(self):
        return self._native.get_used_host_memory()

    def get_available_device_memory(self, device_id):
        return self._native.get_available_device_memory(device_id)

    def get_exception_enabled(self, exception_type):
        if not isinstance(exception_type, ExceptionType):
            return TypeError('exception_type is not of type pyoptix.enums.ExceptionType')

        return self._native.get_exception_enabled(exception_type)

    def set_exception_enabled(self, exception_type, enabled=True):
        if not isinstance(exception_type, ExceptionType):
            return TypeError('exception_type is not of type pyoptix.enums.ExceptionType')

        self._native.set_exception_enabled(exception_type, enabled)

    def get_print_enabled(self):
        return self._native.get_print_enabled()

    def set_print_enabled(self, enabled=True):
        self._native.set_print_enabled(enabled)

    def get_print_buffer_size(self):
        return self._native.get_print_buffer_size()

    def set_print_buffer_size(self, print_buffer_size):
        if not isinstance(print_buffer_size, int) or print_buffer_size < 0:
            raise TypeError('Print buffer size is the number of bytes that must be a positive integer')

        self._native.set_print_buffer_size(print_buffer_size)

    def set_ray_generation_program(self, entry_point_index, program):
        self._native.set_ray_generation_program(entry_point_index, program._native)
        self._ray_gen_programs[entry_point_index] = program

    def get_ray_generation_program(self, entry_point_index):
        return self._ray_gen_programs[entry_point_index]

    def set_exception_program(self, entry_point_index, program):
        self._native.set_exception_program(entry_point_index, program._native)
        self._ray_gen_programs[entry_point_index] = program

    def get_exception_program(self, entry_point_index):
        return self._exception_programs[entry_point_index]

    def set_miss_program(self, ray_type_index, program):
        self._native.set_miss_program(ray_type_index, program._native)
        self._miss_programs[ray_type_index] = program

    def get_miss_program(self, ray_type_index):
        return self._miss_programs[ray_type_index]

    def set_all_exceptions_enabled(self, is_enabled):
        self._native.set_exception_enabled(ExceptionType.all, is_enabled)

    def validate(self):
        self._native.validate()

    def compile(self):
        self._native.compile()


_context_stack = []


def current_context():
    return _context_stack[-1]


def push_context(ctx):
    if not isinstance(ctx, Context):
        raise TypeError('ctx must be an instance of Context')
    _context_stack.append(ctx)


def pop_context():
    return _context_stack.pop()


def _clear_context_stack():
    while len(_context_stack) > 0:
        ctx = _context_stack.pop()
        ctx._mark_all_objects_destroyed()


atexit.register(_clear_context_stack)

# Create a context automatically
Context()

File: driver/context.cpp
#include "context.h"


NativeContextWrapper::NativeContextWrapper(): NativeScopedWrapper() {
    this->context = optix::Context::create();
    this->set_scoped_object(this->context.get());
}

NativeContextWrapper::~NativeContextWrapper() {
    this->context->destroy();
}

unsigned int NativeContextWrapper::get_ray_type_count() {
    return this->context->getRayTypeCount();
}

unsigned int NativeContextWrapper::get_entry_point_count() {
    return this->context->getEntryPointCount();
}

void NativeContextWrapper::set_ray_type_count(unsigned int ray_type_count) {
    this->context->setRayTypeCount(ray_type_count);
}

void NativeContextWrapper::set_entry_point_count(unsigned int entry_point_count) {
    this->context->setEntryPointCount(entry_point_count);
}

void NativeContextWrapper::set_ray_generation_program(unsigned int entry_point_index, NativeProgramWrapper* ray_generation_program) {
    this->context->setRayGenerationProgram(entry_point_index, ray_generation_program->get_native());
}

void NativeContextWrapper::set_exception_program(unsigned int entry_point_index, NativeProgramWrapper* exception_program) {
    this->context->setExceptionProgram(entry_point_index, exception_program->get_native());
}

void NativeContextWrapper::set_miss_program(unsigned int ray_type_index, NativeProgramWrapper* miss_program) {
    this->context->setMissProgram(ray_type_index, miss_program->get_native());
}

void NativeContextWrapper::compile() {
    this->context->compile();
}

void NativeContextWrapper::launch_1d(unsigned int entry_point_index, int width) {
    this->context->launch(entry_point_index, width);
}

void NativeContextWrapper::launch_2d(unsigned int entry_point_index, int width, int height) {
    this->context->launch(entry_point_index, width, height);
}

void NativeContextWrapper::launch_3d(unsigned int entry_point_index, int width, int height, int depth) {
    this->context->launch(entry_point_index, width, height, depth);
}

int NativeContextWrapper::get_cpu_num_of_threads() {
    return context->getCPUNumThreads();
}

void NativeContextWrapper::set_cpu_num_of_threads(int threadCount) {
    context->setCPUNumThreads(threadCount);
}

int NativeContextWrapper::get_stack_size() {
    return context->getStackSize();
}

void NativeContextWrapper::set_stack_size(int stack_size_bytes) {
    context->setStackSize(stack_size_bytes);
}


int NativeContextWrapper::get_available_devices_count() {
    return optix::Context::getDeviceCount();
}

std::string NativeContextWrapper::get_device_name(int device_id) {
    return context->getDeviceName(device_id);
}

boost::python::tuple NativeContextWrapper::get_device_compute_capability(int device_id) {
    int computeCapabilities[2];
    context->getDeviceAttribute(device_id, RT_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY, sizeof(computeCapabilities), computeCapabilities);
    return boost::python::make_tuple(computeCapabilities[0], computeCapabilities[1]);
}

int NativeContextWrapper::get_enabled_device_count() {
    return context->getEnabledDeviceCount();
}

std::vector<int> NativeContextWrapper::get_enabled_devices() {
    return context->getEnabledDevices();
}

void NativeContextWrapper::set_devices(std::vector<int> devices) {
    context->setDevices(devices.begin(), devices.end());
}

unsigned long NativeContextWrapper::get_used_host_memory() {
    return static_cast<unsigned long>(context->getUsedHostMemory());
}

unsigned long NativeContextWrapper::get_available_device_memory(int device_id) {
    return static_cast<unsigned long>(context->getAvailableDeviceMemory(device_id));
}

void NativeContextWrapper::set_exception_enabled(RTexception exception, bool enabled) {
    this->context->setExceptionEnabled(exception, enabled);
}

bool NativeContextWrapper::get_exception_enabled(RTexception exception) {
    return this->context->getExceptionEnabled(exception);
}

void NativeContextWrapper::set_print_enabled(bool enabled){
    this->context->setPrintEnabled(enabled);
}

bool NativeContextWrapper::get_print_enabled() {
    return this->context->getPrintEnabled();
}

void NativeContextWrapper::set_print_buffer_size(int buffer_size_bytes){
    this->context->setPrintBufferSize(buffer_size_bytes);
}

int NativeContextWrapper::get_print_buffer_size() {
    return this->context->getPrintBufferSize();
}

NativeProgramWrapper* NativeContextWrapper::create_program_from_file(std::string file_name, std::string program_name) {
    return new NativeProgramWrapper(this->context->createProgramFromPTXFile(file_name, program_name));
}

NativeBufferWrapper* NativeContextWrapper::create_buffer(int buffer_type) {
    return new NativeBufferWrapper(this->context->createBuffer(buffer_type));
}

NativeTextureSamplerWrapper* NativeContextWrapper::create_texture_sampler() {
    return new NativeTextureSamplerWrapper(this->context->createTextureSampler());
}

NativeGeometryWrapper* NativeContextWrapper::create_geometry() {
    return new NativeGeometryWrapper(this->context->createGeometry());
}

NativeMaterialWrapper* NativeContextWrapper::create_material() {
    return new NativeMaterialWrapper(this->context->createMaterial());
}

NativeGeometryInstanceWrapper* NativeContextWrapper::create_geometry_instance() {
    return new NativeGeometryInstanceWrapper(this->context->createGeometryInstance());
}

NativeGroupWrapper* NativeContextWrapper::create_group() {
    return new NativeGroupWrapper(this->context->createGroup());
}

NativeGeometryGroupWrapper* NativeContextWrapper::create_geometry_group() {
    return new NativeGeometryGroupWrapper(this->context->createGeometryGroup());
}

NativeTransformWrapper* NativeContextWrapper::create_transform() {
    return new NativeTransformWrapper(this->context->createTransform());
}

NativeSelectorWrapper* NativeContextWrapper::create_selector() {
    return new NativeSelectorWrapper(this->context->createSelector());
}

NativeAccelerationWrapper* NativeContextWrapper::create_accelerator(std::string builder, std::string traverser) {
    return new NativeAccelerationWrapper(this->context->createAcceleration(builder.c_str(), traverser.c_str()));
}

void NativeContextWrapper::boost_python_expose() {
    namespace bp = boost::python;

    bp::class_<NativeContextWrapper, bp::bases<NativeScopedWrapper> >(
                "NativeContextWrapper",
                "Wraps optix::Context class",
                bp::init<>())

            .def("get_ray_type_count", &NativeContextWrapper::get_ray_type_count)
            .def("set_ray_type_count", &NativeContextWrapper::set_ray_type_count)
            .def("get_entry_point_count", &NativeContextWrapper::get_entry_point_count)
            .def("set_entry_point_count", &NativeContextWrapper::set_entry_point_count)
            .def("get_cpu_num_of_threads", &NativeContextWrapper::get_cpu_num_of_threads)
            .def("set_cpu_num_of_threads", &NativeContextWrapper::set_cpu_num_of_threads)
            .def("get_stack_size", &NativeContextWrapper::get_stack_size)
            .def("set_stack_size", &NativeContextWrapper::set_stack_size)
            .def("get_available_devices_count", &NativeContextWrapper::get_available_devices_count)
            .def("get_device_name", &NativeContextWrapper::get_device_name)
            .def("get_device_compute_capability", &NativeContextWrapper::get_device_compute_capability)
            .def("get_enabled_device_count", &NativeContextWrapper::get_enabled_device_count)
            .def("get_enabled_devices", &NativeContextWrapper::get_enabled_devices)
            .def("set_devices", &NativeContextWrapper::set_devices)
            .def("get_used_host_memory", &NativeContextWrapper::get_used_host_memory)
            .def("get_available_device_memory", &NativeContextWrapper::get_available_device_memory)
            .def("get_exception_enabled", &NativeContextWrapper::get_exception_enabled)
            .def("set_exception_enabled", &NativeContextWrapper::set_exception_enabled)
            .def("get_print_enabled", &NativeContextWrapper::get_print_enabled)
            .def("set_print_enabled", &NativeContextWrapper::set_print_enabled)
            .def("get_print_buffer_size", &NativeContextWrapper::get_print_buffer_size)
            .def("set_print_buffer_size", &NativeContextWrapper::set_print_buffer_size)
            .def("set_ray_generation_program", &NativeContextWrapper::set_ray_generation_program)
            .def("set_exception_program", &NativeContextWrapper::set_exception_program)
            .def("set_miss_program", &NativeContextWrapper::set_miss_program)
            .def("compile", &NativeContextWrapper::compile)
            .def("create_program_from_file", &NativeContextWrapper::create_program_from_file, bp::return_value_policy<bp::manage_new_object>())
            .def("create_buffer", &NativeContextWrapper::create_buffer, bp::return_value_policy<bp::manage_new_object>())
            .def("create_texture_sampler", &NativeContextWrapper::create_texture_sampler, bp::return_value_policy<bp::manage_new_object>())
            .def("create_geometry", &NativeContextWrapper::create_geometry, bp::return_value_policy<bp::manage_new_object>())
            .def("create_material", &NativeContextWrapper::create_material, bp::return_value_policy<bp::manage_new_object>())
            .def("create_geometry_instance", &NativeContextWrapper::create_geometry_instance, bp::return_value_policy<bp::manage_new_object>())
            .def("create_group", &NativeContextWrapper::create_group, bp::return_value_policy<bp::manage_new_object>())
            .def("create_geometry_group", &NativeContextWrapper::create_geometry_group, bp::return_value_policy<bp::manage_new_object>())
            .def("create_transform", &NativeContextWrapper::create_transform, bp::return_value_policy<bp::manage_new_object>())
            .def("create_selector", &NativeContextWrapper::create_selector, bp::return_value_policy<bp::manage_new_object>())
            .def("create_accelerator", &NativeContextWrapper::create_accelerator, bp::return_value_policy<bp::manage_new_object>())
            .def("launch_1d", &NativeContextWrapper::launch_1d)
            .def("launch_2d", &NativeContextWrapper::launch_2d)
            .def("launch_3d", &NativeContextWrapper::launch_3d);
}
